### PR TITLE
Fix constant propagation in builtins and UserClasses

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118994

Fixes https://github.com/pytorch/pytorch/issues/118675

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py [hunk 1]
```diff
@@ -1289,12 +1291,13 @@ def _grad_changed(old, new):
                 if example_value.grad is not None:
                     unimplemented("getattr on non-None grad - NYI")
                 return ConstantVariable(None)
+        elif obj.is_python_constant() and hasattr(obj.as_python_constant(), name):
+            return builder(getattr(obj.as_python_constant(), name))
```
[the remark this answers is not in the document]
Comment (lezcano) [reply torch/_dynamo/variables/builtin.py:1295]: Note that `is_python_constant` is much more general than `ConstantVariable`. That's one of the reasons why,e ven after fixing the SourcelessBuilder errata, I'm still getting tons of errors.
Let me put this as WIP for now.

[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/_dynamo/variables/builtin.py:1295]: Note that the code I linked is in the `VariableTracker` base class, not `ConstantVariable`. I'm actually surprised that `const_getattr` doesn't have a base implementation. I would think this is pretty general:
```python
return getattr(self.as_python_constant(), name)
```

[the remark this answers is not in the document]
Comment (lezcano) [reply torch/_dynamo/variables/builtin.py]: Implementing this in the base class took me down a rabbit hole, but if I manage to get it right, I think we should get quite a bit of functionality for free. Let's see...

File: torch/_dynamo/variables/tensor.py [hunk 2]
```diff
@@ -190,23 +190,23 @@ def dynamic_getattr(self, tx, name):
             # Which is incorrect, and violates the invariant that all sources should be eval()-able against the scope.
             _input_associated_real_value = eval(self.source.name(), scope)
         except Exception as exc:
-            raise NotImplementedError() from exc
+            return
```
Comment (jansel) [inline torch/_dynamo/variables/tensor.py:193]: Returning `None` seems wrong since the call above is expecting to catch NotImplementedError.

**Review by jansel — CHANGES_REQUESTED** [review]

**Pull request closed without merging**